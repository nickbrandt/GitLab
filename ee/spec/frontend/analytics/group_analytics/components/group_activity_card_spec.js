import { mount } from '@vue/test-utils';
import MockAdapter from 'axios-mock-adapter';
import GroupActivityCard from 'ee/analytics/group_analytics/components/group_activity_card.vue';
import Api from 'ee/api';
import waitForPromises from 'helpers/wait_for_promises';
import MetricCard from '~/analytics/shared/components/metric_card.vue';
import axios from '~/lib/utils/axios_utils';

const TEST_GROUP_ID = 'gitlab-org';
const TEST_GROUP_NAME = 'Gitlab Org';
const TEST_MERGE_REQUESTS_COUNT = { data: { merge_requests_count: 10 } };
const TEST_ISSUES_COUNT = { data: { issues_count: 20 } };
const TEST_NEW_MEMBERS_COUNT = { data: { new_members_count: 30 } };

describe('GroupActivity component', () => {
  let wrapper;
  let mock;

  const createComponent = () => {
    wrapper = mount(GroupActivityCard, {
      provide: {
        groupFullPath: TEST_GROUP_ID,
        groupName: TEST_GROUP_NAME,
      },
    });
  };

  beforeEach(() => {
    mock = new MockAdapter(axios);

    jest
      .spyOn(Api, 'groupActivityMergeRequestsCount')
      .mockReturnValue(Promise.resolve(TEST_MERGE_REQUESTS_COUNT));

    jest.spyOn(Api, 'groupActivityIssuesCount').mockReturnValue(Promise.resolve(TEST_ISSUES_COUNT));

    jest
      .spyOn(Api, 'groupActivityNewMembersCount')
      .mockReturnValue(Promise.resolve(TEST_NEW_MEMBERS_COUNT));
  });

  afterEach(() => {
    wrapper.destroy();
    mock.restore();
  });

  const findMetricCard = () => wrapper.find(MetricCard);

  it('matches the snapshot', () => {
    createComponent();

    return wrapper.vm
      .$nextTick()
      .then(waitForPromises)
      .then(() => {
        expect(wrapper.element).toMatchSnapshot();
      });
  });

  it('fetches MR and issue count and updates isLoading properly', () => {
    createComponent();

    expect(wrapper.vm.isLoading).toBe(true);

    return wrapper.vm
      .$nextTick()
      .then(() => {
        expect(Api.groupActivityMergeRequestsCount).toHaveBeenCalledWith(TEST_GROUP_ID);
        expect(Api.groupActivityIssuesCount).toHaveBeenCalledWith(TEST_GROUP_ID);
        expect(Api.groupActivityNewMembersCount).toHaveBeenCalledWith(TEST_GROUP_ID);

        waitForPromises();
      })
      .then(() => {
        expect(wrapper.vm.isLoading).toBe(false);
        expect(wrapper.vm.metrics.mergeRequests.value).toBe(10);
        expect(wrapper.vm.metrics.issues.value).toBe(20);
        expect(wrapper.vm.metrics.newMembers.value).toBe(30);
      });
  });

  it('passes the metrics array to the metric card', () => {
    createComponent();

    return wrapper.vm
      .$nextTick()
      .then(waitForPromises)
      .then(() => {
        expect(findMetricCard().props('metrics')).toEqual([
          { key: 'mergeRequests', value: 10, label: 'Merge requests opened' },
          { key: 'issues', value: 20, label: 'Issues opened' },
          { key: 'newMembers', value: 30, label: 'Members added' },
        ]);
      });
  });
});
