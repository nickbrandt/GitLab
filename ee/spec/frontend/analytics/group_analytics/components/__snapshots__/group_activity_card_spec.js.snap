// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupActivity component matches the snapshot 1`] = `
<div
  class="gl-card gl-mb-5"
>
  <div
    class="gl-card-header"
  >
    <strong>
      Recent activity (last 90 days)
    </strong>
  </div>
   
  <div
    class="gl-card-body"
  >
    <div
      class="gl-display-flex"
    >
      <div
        class="js-metric-card-item gl-flex-grow-1 gl-text-center"
      >
        <h3
          class="gl-my-2"
        >
          10
        </h3>
         
        <p
          class="text-secondary gl-font-sm gl-mb-2"
        >
          
          Merge requests opened
          
          <!---->
        </p>
      </div>
      <div
        class="js-metric-card-item gl-flex-grow-1 gl-text-center"
      >
        <h3
          class="gl-my-2"
        >
          20
        </h3>
         
        <p
          class="text-secondary gl-font-sm gl-mb-2"
        >
          
          Issues opened
          
          <!---->
        </p>
      </div>
      <div
        class="js-metric-card-item gl-flex-grow-1 gl-text-center"
      >
        <h3
          class="gl-my-2"
        >
          30
        </h3>
         
        <p
          class="text-secondary gl-font-sm gl-mb-2"
        >
          
          Members added
          
          <!---->
        </p>
      </div>
    </div>
  </div>
   
  <!---->
</div>
`;
